Incorrect suggestion when trying to write to an immutable field in an async function

Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=df6e36b97b73bd3e3ea0fdfe9325b1c9

```rust
struct S { foo: usize }
impl S {
    async fn bar(&self) {
        self.foo += 1;
    }
}
```

The current output is:

```
error[E0594]: cannot assign to `self.foo`, which is behind a `&` reference
 --> src/lib.rs:4:9
  |
3 |     async fn bar(&self) {
  |                  ----- help: consider changing this to be a mutable reference: `&mut S`
4 |         self.foo += 1;
  |         ^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be written
```

<!-- The following is not always necessary. -->
Ideally the output should look like:

```
error[E0594]: cannot assign to `self.foo`, which is behind a `&` reference
 --> src/lib.rs:4:9
  |
3 |     async fn bar(&self) {
  |                  ----- help: consider changing this to be a mutable reference: `&mut self`
4 |         self.foo += 1;
  |         ^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be written
```

<!--
If the problem is not self-explanatory, please provide a rationale for the
change.
-->

`&mut S` is not a valid argument; it needs to be either `&mut self` or `some_name: &mut S`. For some reason this problem only happens when the function is async.

Another similar error, not related to async this time:
```
error[E0596]: cannot borrow `*callback` as mutable, as it is behind a `&` reference
   --> boring/src/ssl/callbacks.rs:278:5
    |
273 |     let callback = ctx
    |         -------- help: consider changing this to be a mutable reference: `&mut F`
...
278 |     callback(ctx, session)
    |     ^^^^^^^^ `callback` is a `&` reference, so the data it refers to cannot be borrowed as mutable
```
`let &mut F = ...` isn't valid Rust.

--- a/compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs
@@ -488,12 +488,32 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                                     // don't create labels for compiler-generated spans
                                     Some(_) => None,
                                     None => {
-                                        let (span, suggestion) = suggest_ampmut(
-                                            self.infcx.tcx,
-                                            local_decl,
-                                            opt_assignment_rhs_span,
-                                            *opt_ty_info,
-                                        );
+                                        let (span, suggestion) = if name != kw::SelfLower {
+                                            suggest_ampmut(
+                                                self.infcx.tcx,
+                                                local_decl,
+                                                opt_assignment_rhs_span,
+                                                *opt_ty_info,
+                                            )
+                                        } else {
+                                            match local_decl.local_info.as_deref() {
+                                                Some(LocalInfo::User(ClearCrossCrate::Set(
+                                                    mir::BindingForm::Var(mir::VarBindingForm {
+                                                        opt_ty_info: None,
+                                                        ..
+                                                    }),
+                                                ))) => {
+                                                    suggest_ampmut_self(self.infcx.tcx, local_decl)
+                                                }
+                                                // explicit self (eg `self: &'a Self`)
+                                                _ => suggest_ampmut(
+                                                    self.infcx.tcx,
+                                                    local_decl,
+                                                    opt_assignment_rhs_span,
+                                                    *opt_ty_info,
+                                                ),
+                                            }
+                                        };
                                         Some((true, span, suggestion))
                                     }
                                 }

--- a/src/test/ui/borrowck/issue-93093.rs
+++ b/src/test/ui/borrowck/issue-93093.rs
@@ -1,0 +1,14 @@
+// edition:2018
+struct S {
+    foo: usize,
+}
+impl S {
+    async fn bar(&self) { //~ HELP consider changing this to be a mutable reference
+        //~| SUGGESTION &mut self
+        self.foo += 1; //~ ERROR cannot assign to `self.foo`, which is behind a `&` reference [E0594]
+    }
+}
+
+fn main() {
+    S { foo: 1 }.bar();
+}

--- a/src/test/ui/borrowck/issue-93093.stderr
+++ b/src/test/ui/borrowck/issue-93093.stderr
@@ -1,0 +1,12 @@
+error[E0594]: cannot assign to `self.foo`, which is behind a `&` reference
+  --> $DIR/issue-93093.rs:8:9
+   |
+LL |     async fn bar(&self) {
+   |                  ----- help: consider changing this to be a mutable reference: `&mut self`
+LL |
+LL |         self.foo += 1;
+   |         ^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be written
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0594`.

